# SLM Bioinformatics: Gene expression analysis
This set of reference data and a commenting manual can be used to support the self-learning of bioinformatic applications. It teaches the practical use of next-generation sequencing data to investigate gene regulation at the transcriptome level. Here we provide two self-learning modules:

1. Differential gene transcriptional analysis between two A. halleri accessions differing in their Cadmium tolerance (based on [Lee et al. 2017](https://onlinelibrary.wiley.com/doi/full/10.1111/tpj.15544))
* To successfully run this self-learning module please download
  * [Data](https://ruhr-uni-bochum.sciebo.de/s/TeeDqNyqAjAgJjW))
  * [Additional software](https://ruhr-uni-bochum.sciebo.de/s/opCaPMoRF3nlvun))
* UPDATE expected by November 2023 including:
  * New A. halleri Lan3.1 genome
  * Updated Programs e.g. cutadapt, hisat2, feature_counts etc.
  * NEW recirpocal best BLAST / one-way best BLAST based orthology including copy number expanded genes

2. Differential gene transcriptional analysis in A. thaliana under ambient and heat stress conditions (based on [Pietzenuk et al. 2017](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-1072-3))  
* CURRENTLY UNDER CONSTRUCTION! COME BACK OCTOBER 13th 2023
* If you are interested in TE analysis, consider the heat stress data and use the sripts from [COMEX](https://github.com/bpietzenuk/COMEX)
